### ISSUE TITLE
Incorrect sequence read when loading ab1 file with low quality by SeqIO.read

### ISSUE THREAD

**Issue by mrdzne** [biopython/biopython#1531]
Hey there,

I am reporting a problem with Biopython 1.70, Python version 3.6.4 on Linux-4.15.3-1-ARCH-x86_64-with-arch.

It appears that the nucleotide sequence is incorrectly parsed, when loading the attached file test.ab1 with SeqIO.read. Specifically, nucleotide 336 should either be a G or a N, not a C. The sequencing quality is bad at this position, so probably it should be a N.

[seq.zip](https://github.com/biopython/biopython/files/1728132/seq.zip)

**Comment by peterjc** [comment]
Thanks for providing a test case, that should make things much easier to investigate.

**Comment by mrdzne** [comment]
`#!/usr/bin/python
from Bio import SeqIO

record=SeqIO.read("test.ab1","abi")
print("Read at position 336: " + record.seq[335 ])
print("Quality of sequencing at position 336: " + str(record.letter_annotations['phred_quality'][335]))`

See the above minimal working example. 

If you open that file manually in SnapGene or SequenceViewer (these are the two Windows freeware ab1 file readers I checked), the sequence reads a G at this position.

I really do not know enough about the ab1 file format to judge whether Biopython is right or wrong on this one. I opened the bug here though because its read is discordant to two other readers.

**Comment by MarkusPiotrowski** [comment]
I opened the file in another freeware to read abi files (Chromas) where you can display the "original" sequence, and this has a C. So this is not a real bug. The file seems to contain two sequences, the sequence called by the base caller and the user edited sequence. It seems that SeqIO returns the first one.

**Comment by mrdzne** [comment]
I agree that the sequence is ambiguous (hence I think N is maybe more appropriate).
But also in Chromas at position 336 the peak for G is higher than the one for C, right?

I doubt the file has been manually edited though - it came from the sequencing provider like this. I will re-check with my coworker tomorrow though.

**Comment by bow** [comment]
@mrdzne, thanks for reporting the error and including the test case. @MarkusPiotrowski is right, this isn't a bug per-se, but a decision that we made earlier to show the sequence as called by the base caller.

If you're interested, the details are in [the ABI file specification](http://www6.appliedbiosystems.com/support/software_community/ABIF_File_Format.pdf), page 36. Look for the tag called `PBAS` and there you can see that there are two variants of the tag. The first variant, as @MarkusPiotrowski mentioned, stores the user-edited sequence while the other shows the sequence as called by the basecaller.

@peterjc Do you think it's a good idea to expose the option of which `PBAS` variant to use? We'd still use the current one as a default, but I have to admit that was a deliberate choice on my end that may not suite everyone.

**Comment by MarkusPiotrowski** [comment]
@mrdzne Keep in mind that  ```SeqIO``` is just reporting the sequence as it was stored in the ab1 file, it is not re-interpreting the chromatogram data. And I also think that it is quite usual in sequencing facilities that _they_ edit the original sequence. As a wet bench biologist I agree with you that the interpretation should be 'S' (G or C). Could be a mixture of two sequences with a single nucleotide exchange, e.g. a PCR of a gene with two different alleles (e.g. a heterozygous mutant).

@bow Since most other software seem to report the user-edited sequence, I think returning the user-edited sequence may be more appropriate.

**Comment by mrdzne** [comment]
@MarkusPiotrowski @bow you guys were right, the sequence was user-edited by us at a previous stage. Probably the plasmid is a mixture of two variants, as resequencing yielded the same ambiguity at this position at a comparable quality.

I think an option to fetch the user-edited sequence would be great. Anyway, thanks for this great piece of software!

**Comment by peterjc** [comment]
An option to prefer the user-edited sequence seems a good idea (optional argument to the ABI parser), or perhaps even as a named format variant in the SeqIO interface?

Would ``abi`` (current default) and ``abi-edited`` make sense (use edited sequence if present)?

**Comment by bow** [comment]
@peterjc , if possible I'd like to expose this via `SeqIO.[read|parse]` as a keyword argument that is ignored or will raise an error for non-ABI file formats.

Making a new `abi-edited` format works, but then there would be no option for users to use the user-edited variant with trimming (`abi-trim`), for example. Of course we can also add `abi-edited-trim`, but I think a better way is to have something like `abi_trim=[True|False]` and `abi_edited=[True|False]` as keyword arguments. This way, we can keep the format names small, but allow more user-defined configuration. We can also extend this behavior to other formats, as required.

We can have the current behavior as the default values for these (so `abi_trim=False` and `abi_edited=False`) to keep backwards compatibility.

**Comment by peterjc** [comment]
I've thus far resisted allowing arbitrary keyword arguments to ``SeqIO`` (*Special cases aren't special enough to break the rules*), and the ``format-variant`` approach has generally sufficed. If you are suggesting allowing only keyword arguments with a format-underscore prefix, that gets more interesting, as it side steps the question about consistency in common argument names like (like trimming, or for output wrapping options). Interesting... maybe we should prototype that?

This aside, adding an edited argument to the underlying ``AbiIterator`` function seems absolutely fine:
https://github.com/biopython/biopython/blob/biopython-170/Bio/SeqIO/AbiIO.py#L333

**Comment by bow** [comment]
@peterjc A prototype sounds nice. I'll try to write something up soon and ping everyone ~
